### DISCUSSION TITLE
floating point decimal

### DISCUSSION THREAD

**Discussion by RafieeAshkan** [meta-pytorch/botorch#1086] [Q&A]
Hi
Sometimes, the acquisition function optimisation recommend sample that are very close (differ only in high decimal values), for example these two new recommended solutions 
```
0.0,0.819989962452467,0.47227432081417425,0.5890579273067887,0.9944247468980589,0.9848942568382668
0.0,0.8199899525564098,0.4722743157909559,0.5890579317025663,0.994424729005317,0.9848942553202004
```
in most cases such accuracy is not practical.  Is it possible to pass a rounding argument to "optimize_acqf" to get more different points?
Many thanks

**Comment by Balandat** [comment]
I guess I'd like to understand the larger context here. Am I understanding correctly that the issue is not that the parameters are in fact discrete and that you want to round, but that the acquisition function optimization provides suggestions that are essentially identical to training observations?

I that is the case then a few comments / questions:
- What acquisition function are you using? 
- Are your observations noisy? What kind of model are you using? Are you inferring the noise level (e.g. the default `SingleTaskGP`) or are you passing variances (e.g. with a `FixedNoiseGP`)? If there is indeed observation noise, then re-suggesting the same point may not be a problem as repeated observations at the same point will help better estimate the latent function.
- There may be other domain-specific considerations for why you'd want to avoid evaluating the points close to each other. But if that's the case, I think the right place to encourage that would be in the acquisition function itself, rather than by doing ad-hoc rounding ex-post. If that's something you'd like to try one quick and dirty way could be to use a [`PenalizedAcquisitionFunction`](https://github.com/pytorch/botorch/blob/main/botorch/acquisition/penalized.py#L1420) - you could implement your own penalty term that just adds a repelling penalty around previously evaluated points. The tricky part there will be how to get the scale right, but maybe you can fiddle around a bit - a more principled approach might require implementing a new acquisition function.

**Comment by RafieeAshkan** [comment]
Thanks for the quick respond @Balandat
To give more info, the observations are coming from a fluid dynamics simulations. I am using qEHVI acquisition function with 2 objectives and 2 constraints and the GP model is SingleTaskGP. Since the evaluations can be very expensive (needs to run multiple CFD simulations on an HPC cluster) and therefore prefer to avoid estimations of points that are inherently the same from design point of view.

**Reply by Balandat** [reply]
Hmm interesting. Multiple objectives and multiple constraints could make for a tough problem. I wonder if the constraints are really tight and the acquisition function has a hard time finding feasible points. Do you have some data that you could share for us to see what might be going on here? 

cc @sdaulton

**Reply by RafieeAshkan** [reply]
certainly.
Here are the objective values (negated as I am doing a minimisation problem) from first initial random samples. The last value is essentially one of the simulations was failed. As the failure rate is generally very small and random, I drop them in model fitting

```
-37139.5766115,-139910.34617
-37801.58595,-128966.548705
-38746.8158535,-154875.88792
-42974.5665795,-143565.088815
-47761.9458895,-146465.32003499998
-42908.20599,-141188.08175
-29772.4781805,-188829.70078
-47359.281178000005,-185715.2472
-36388.980457,-127764.80679
-38001.001263,-117681.30741
-38330.4751345,-128064.54573
-35927.062094,-130910.305475
-55164.6709915,-172317.105875
-60000.0,-145400.11156
```

associated to these objectives, these are the optimisation variables (in this case I have 6 design variables), here the values are in [0,1]

```
0.1994731277227401,0.1709322035312652,0.7493569254875183,0.1843137741088867,0.5505224466323853,0.7656387686729431
0.3626673817634582,0.9827781915664672,0.9238153100013732,0.1526303440332412,0.7649058103561401,0.4879069924354553
0.2037077248096466,0.4697332084178924,0.7626118063926697,0.787773072719574,0.2797711193561554,0.0287332758307456
0.3490467667579651,0.0905924141407013,0.7045403718948364,0.0396272838115692,0.2078758329153061,0.6575039625167847
0.8252790570259094,0.7682657241821289,0.1446177810430526,0.4590553045272827,0.456321120262146,0.9306137561798096
0.3396573662757873,0.2267412990331649,0.2115753144025802,0.186450719833374,0.9086307883262634,0.1065155565738678
0.037636674940586,0.1367794573307037,0.5614128112792969,0.2773758471012115,0.2019493132829666,0.1012429445981979
0.4443546235561371,0.3869217932224273,0.803887665271759,0.4513152241706848,0.2990014255046844,0.024661686271429
0.1114875376224517,0.3511012494564056,0.4846924543380737,0.913844347000122,0.7792673110961914,0.7379682660102844
0.6619455814361572,0.6836562156677246,0.6333109736442566,0.7773427963256836,0.6300293803215027,0.7341881394386292
0.4384100437164306,0.2038832902908325,0.2436946332454681,0.3034232854843139,0.5469076633453369,0.5426182150840759
0.3404804170131683,0.882500946521759,0.3839507102966308,0.7234951853752136,0.3293978571891784,0.7795841693878174
0.5547549724578857,0.967616856098175,0.96878844499588,0.6907500624656677,0.0068457815796136,0.0290455259382724
0.9301573038101196,0.4249435365200043,0.0701839178800582,0.0192474909126758,0.4301266372203827,0.2385824918746948
```

In addition, here are the constraints (negative being feasible)

```
-1.91,-0.5983300056002951
-1.837,-1.2430417699889469
-2.253,-0.3557395094295632
-2.401,0.05305188219869
-2.1495,-1.911160965015409
-2.012,-0.2255078822264611
-2.272,1.4465103962167554
-2.45,-0.0898833737022091
-1.653,-1.3724932987355318
-1.987,-1.9831012006705624
-2.127,-0.7292980940540996
-2.058,-1.6736189582999823
-2.904,-0.6848632741461973
-2.71,-0.4827659254009546
```

Here are the new recommended sample (using q=4):

```
0.0,0.819989962452467,0.47227432081417425,0.5890579273067887,0.9944247468980589,0.9848942568382668
0.0,0.8199899525564098,0.4722743157909559,0.5890579317025663,0.994424729005317,0.9848942553202004
0.027498212304091673,0.9769163378044284,0.6351962335668737,0.8104741896452122,1.0,1.0
3.0071876721574534e-5,0.7793765535941397,0.5348468513171332,0.8867551063142654,0.6947080179802572,0.997605002541196
```

I am aiming to do as low number of simulations as possible, so the initial samples follows the recommended "2*(d+1)".
